src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + _preCount + _appCount : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index = 0;
+                for (int i = -_preCount; i != 0; ++i, ++index)
+                {
+                    array[index] = _prepended[i];
+                }
+
+                Debug.Assert(index == _preCount);
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                Debug.Assert(index == array.Length - _appCount);
+                for (int i = 0; i != _appCount; ++i, ++index)
+                {
+                    array[index] = _appended[i];
+                }
+
+                Debug.Assert(index == array.Length);
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>(Math.Max(_appCount + _preCount, 4)) : new List<TSource>(count);
+                for (int i = -_preCount; i != 0; ++i)
+                {
+                    list.Add(_prepended[i]);
+                }
+
+                list.AddRange(_source);
+
+                for (int i = 0; i != _appCount; ++i)
+                {
+                    list.Add(_appended[i]);
+                }
+
+                return list;
+            }
+        }
+    }
+
+    internal class SharedCollection<T>
+    {
+        // Limit the number of elements stored, because the collection being shared makes it possible
+        // for a series of operations to grow the collection large, but for only one of the earlier
+        // operations to be keeping it alive.
+        private const int MaxLength = 64;
+        private T[] _store;
+        private int _count;
+
+        public SharedCollection(T item)
+        {
+            _store = new[] { item };
+            _count = 1;
+        }
+
+        public SharedCollection(T first, T second)
+        {
+            _store = new[] { first, second };
+            _count = 2;
+        }
+
+        public T this[int index] => _store[index >= 0 ? index : index + _store.Length];
+
+        public bool Append(T item, int number)
+        {
+            if (number != _count | number >= MaxLength)
+            {
+                return false;
+            }
+
+            T[] store = _store;
+
+            if (store.Length == number)
+            {
+                T[] newStore = new T[number << 1];
```

Comment:
Another place where I'm just showing my age :wink: I expect it isn't terribly obscure though; there are at least a few places with the same thing in CoreFX and CoreCLR. (I think it doesn't normally do that optimisation, btw, and certainly there are differences between shifting and multiplication that won't come up here but do man it can only be considered equivalent in certain places. That said, while it may be my habit to bit-shift in cases like this, I don't think it's a habit worth acquiring).